lib/ecto/query/planner.ex

```diff
@@ -938,17 +938,14 @@ defmodule Ecto.Query.Planner do
 
   # Expression handling
 
-  defp collect_fields({agg, _, [{{:., _, [{:&, _, [ix]}, field]}, _, []} | _]} = expr,
-                      fields, from, %{select: select} = query, _take)
+  defp collect_fields({agg, _, [field | _]} = expr, fields, from, query, _take)
        when agg in ~w(count avg min max sum)a do
-    type =
-      # TODO: Support the :number type
-      case agg do
-        :count -> :integer
-        :avg -> :any
-        :sum -> :any
-        _ -> source_type!(:select, query, select, ix, field)
-      end
+    {{:value, aggregate_type(agg, field, query)}, [expr | fields], from}
+  end
+
+  defp collect_fields({:filter, _, [{agg, _, [field | _]}, _]} = expr, fields, from, query, _take)
```

Comment:
I would prefer to make it recursive so it always works regardless of changes elsewhere. Something like:

    defp collect_fields({:filter, _, [call, _]} = expr, fields, from, query, take) do
      {type, _, _} = collect_fields(call, fields, from, query, take)
      {type, [expr | fields], from}
    end

Reply:
Yeah, much simpler. I'll replace the original `collect_fields` for aggregates as well, so there will be less change.